models/CouponModel.php

```diff
@@ -851,6 +851,12 @@ public function deduct_coupon_discount( $regular_price, $discount_type, $discoun
 	 * @return bool
 	 */
 	public function is_coupon_valid( object $coupon ): bool {
+		global $coupon_apply_error_msg;
```

Comment:
This method check status, coupon validity (expiry) and usage limit.
Instead of only check status. We need to set error depends on return result
```php
$is_valid = self::STATUS_ACTIVE === $coupon->coupon_status && $this->has_coupon_validity( $coupon ) && $this->has_user_usage_limit( $coupon, get_current_user_id() );

if( ! $is_valid ) {
   $coupon_apply_error_msg = $this->get_coupon_failed_error_msg( 'invalid' );
}

return $is_valid;
```

Reply:
Why shall we check coupon validity & usage limit if the coupon itself is inactive?